GVFS/GVFS.UnitTests.Windows/Windows/Upgrader/ProductUpgraderTests.cs

```diff
@@ -80,6 +80,42 @@ public void UpgradeAvailableOnSlowWhileOnLocalFastRing()
                 expectedUpgradeVersion:UpgradeTests.NewerThanLocalVersion);
         }
 
+        [TestCase]
+        public void RingInNugetFeedURLOverridesUpgradeRing()
+        {
+            Version newVersion;
+            string error;
+            this.SetUpgradeRing(GitHubUpgrader.GitHubUpgraderConfig.RingType.Fast.ToString());
+            this.Upgrader.PretendNewReleaseAvailableAtRemote(UpgradeTests.NewerThanLocalVersion, GitHubUpgrader.GitHubUpgraderConfig.RingType.Fast);
+
+            this.Upgrader.TryQueryNewestVersion(out newVersion, out error).ShouldBeTrue();
+            newVersion.ShouldNotBeNull();
+
+            string feedUrlWithSlowRing = "https://1essharedassets.pkgs.visualstudio.com/_packaging/GVFS@Slow/nuget/v3/index.json";
+            this.LocalConfig.TrySetConfig("upgrade.feedurl", feedUrlWithSlowRing, out error);
+            this.Upgrader.Config.TryLoad(out error).ShouldBeTrue();
+
+            this.Upgrader.TryQueryNewestVersion(out newVersion, out error).ShouldBeTrue();
+            newVersion.ShouldBeNull();

```

Comment:
What in this test is validating that the Feed URL has overridden the upgrade ring?

I can see that the test is validating that `TryQueryNewestVersion` returns true, but how do we know it's using the right ring?

Reply:
line no. 88 sets `upgrade.ring` to `Fast`. Line no. 89 to 92 sets an upgrade available in `Fast`ring  and verifies that Upgrader is able to query and return the new version. 

Now without changing anything (upgrade is still available in `Fast`), line no. 95 sets `upgrade.feedurl` with `Slow` ring config. When we query for newest version, upgrader now returns `Null` in line no. 98. Meaning the `Fast` `upgrade.ring` config got overridden by the `Slow` in `upgrade.feedurl`. We didn't set any upgrade to be available in `Slow`, so upgrader returned `Null` this time for new version.

Will add some comments explaining this behavior in the UT.

Reply:
Updated the UT with comments to clarify the override behavior.

Reply:
> line no. 88 sets `upgrade.ring` to `Fast`. Line no. 89 to 92 sets an upgrade available in `Fast`ring and verifies that Upgrader is able to query and return the new version.

The `Setup` method also makes an update available in the Slow ring:

```
        public virtual void Setup()
        {
            this.Tracer = new MockTracer();
            this.Output = new MockTextWriter();
            this.PrerunChecker = new MockInstallerPrerunChecker(this.Tracer);
            this.LocalConfig = new MockLocalGVFSConfig();

            this.Upgrader = new MockGitHubUpgrader(
                LocalGVFSVersion,
                this.Tracer,
                new GitHubUpgrader.GitHubUpgraderConfig(this.Tracer, this.LocalConfig));

            this.PrerunChecker.Reset();
            this.Upgrader.PretendNewReleaseAvailableAtRemote(
                upgradeVersion: NewerThanLocalVersion,
                remoteRing: GitHubUpgrader.GitHubUpgraderConfig.RingType.Slow);
            this.SetUpgradeRing("Slow");
        }
```

Line 92 (as best I can tell) is only validating that there was *some* newer version returned.  How is the test validating that it's the fast version (and not the slow version)?

> We didn't set any upgrade to be available in `Slow`, so upgrader returned `Null` this time for new version.

See code in `Setup`, it looks like something is available in `Slow`, why is this not getting picked up?  Did it get wiped out when we set it to fast?

Reply:
As discussed offline, added a comment in the UT to clarify the behavior of `this.Upgrader.PretendNewReleaseAvailableAtRemote`.

The flow:
1. `UpgradeTests.Setup()` sets the ring and a pretend-release applicable to most UT cases using `PretendNewReleaseAvailableAtRemote`.
2. When an individual UT calls `PretendNewReleaseAvailableAtRemote` again, it replaces (rather than adding to) what was set by `UpgradeTests.Setup()` method with new Release-Ring info that the UT wants to use.